baseplate/lib/secrets.py

```diff
@@ -464,6 +467,87 @@ def _get_file_data(self, filename: str) -> Tuple[Dict, float]:
         return result
 
 
+class VaultCSIEntry(NamedTuple):
+    mtime: float
+    data: Any
+
+
+class VaultCSISecretsStore(SecretsStore):
+    """Access to secret tokens using a vault CSI mount with automatic refresh.
+
+    This store allows access to secrets stored in vault through the CSI interface.
+    It performs caching and will automatically reload secrets from disk. Generally
+    do not cache or store the values returned by this class's methods, rather get
+    them from this class each time you need them. The secrets are served from memory
+    when possible, so there's little performance impact in doing so, and you will be
+    sure to always have the current version in the face of key rotation etc.
+    """
+
+    path: Path
+    data_symlink: Path
+    cache: Dict[str, VaultCSIEntry]
+
+    def __init__(
+        self,
+        path: str,
+        parser: SecretParser,
+    ):  # pylint: disable=super-init-not-called
+        self.path = Path(path)
+        self.parser = parser
+        self.cache = {}
+        self.data_symlink = self.path.joinpath("..data")
+        if not self.path.is_dir():
+            raise ValueError(f"Expected {self.path} to be a directory.")
+        if not self.data_symlink.is_dir():
+            raise ValueError(
+                f"Expected {self.data_symlink} to be a directory. Verify {self.path} is the root of the Vault CSI mount."
+            )
+
+    def get_vault_url(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def get_vault_token(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def _get_mtime(self) -> float:
+        """Modification time is store-wide for CSI secrets."""
+        return os.path.getmtime(self.data_symlink)
+
+    def _raw_secret(self, name: str) -> Any:
+        try:
+            with open(self.data_symlink.joinpath(name), "r", encoding="UTF-8") as fp:
+                return self.parser(json.load(fp))
+        except FileNotFoundError:
+            # Every few minutes, the Vault CSI will create a new directory and swap over to it.
+            # There is a race condition where we could resolve the symlink and then the file could
+            # There is a race condition where we could resolve the symlink and then the file could
+            # be deleted. Given the timeframe between such deletions, a single retry is sufficient
+            # to double-check whether the FileNotFoundError was due to this race or an
+            # actually-nonexistent path.
```

Comment:
Can this actually happen? I would expect this to make an `open` syscall on the non-resolved symlink (e.g. `open("<csi_root>/..data/secret")` since `joinpath` doesn't resolve symlinks) which the kernel would handle resolving. Once you get an open file handle it shouldn't matter if it gets deleted afterwards, right?

I tried to roughly test this by running this in one terminal:
```
while :; do
    mkdir -p a
    echo foo > a/secret
    ln -s a ..data-new
    mv -T ..data-new ..data
    rm -rf b
    mkdir -p b
    echo bar > b/secret
    ln -s b ..data-new
    mv -T ..data-new ..data
    rm -rf a
done
```

...and this in another:
```
python3.12 -c 'while True: print(open("..data/secret").read())'
```

and I can never reproduce any issues, even if I add an arbitrary delay between opening the file and reading its contents. (If I have it resolve the path first, there are issues frequently.)

Reply:
This was all theoretical. I never tried to produce the issue